ENH: sparse.csgraph.laplacian: add `signed_graph_variant` parameter and support 'opposing' and 'unsigned'

math-hiyoko opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Closes #22685 


#### What does this implement/fix?
This PR adds a new `signed_graph_variant` option to `scipy.sparse.csgraph.laplacian` to improve support for signed networks (signed_graph_variant='standard').

- `signed_graph_variant='standard'` is standard degree calculation using edge weights.
- `signed_graph_variant='opposing'` computes the degree using absolute value only for the degree computation, so that positive and negative weights cannot cancel out.
- `signed_graph_variant='unsigned'` computes both degree and adjacency using absolute value, discarding all sign information.

This change follows the proposal in #22685 and is based on the approach discussed in [Kunegis et al. (2010)](https://www.researchgate.net/publication/220906703_Spectral_Analysis_of_Signed_Graphs_for_Clustering_Prediction_and_Visualization)
Detailed discussions on the reasoning behind each type of variant can be found in [Shi et al. (2019)](https://doi.org/10.1137/17M1134172)

The default behavior (`signed_graph_variant='standard'`) remains unchanged to preserve backward compatibility.

#### Additional information
When both `signed_graph_variant='opposing'` and `symmetrized=True`, the symmetrization is applied *after* taking the absolute value of the input matrix. This ensures that the computed degrees reflect the magnitude of all connections symmetrically, which is particularly relevant for signed adjacency matrices where negative values represent inhibitory relationships.

math-hiyoko committed:
ENH: sparse.csgraph.laplacian support use_abs

math-hiyoko committed:
FIX: the case csgraph is np.ndarray

j-bowhay commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -20,6 +20,7 @@ def laplacian(
     form="array",
     dtype=None,
     symmetrized=False,
+    use_abs=False,
```
Just a drive by comment: if we're adding a new argument, let's make it keyword only
```suggestion
    *,
    use_abs=False,
```

math-hiyoko committed:
Update scipy/sparse/csgraph/_laplacian.py

make use_abs keyword only

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

math-hiyoko committed:
FIX: benchmarktest

math-hiyoko committed:
FIX: function behavior

math-hiyoko committed:
FIX: test according to function behavior

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko commented:
Everything is ready now

lucascolley commented:
@fernandodiazdiaz would you like to take a look?

fernandodiazdiaz commented:
> @fernandodiazdiaz would you like to take a look?

Looks perfect to me!

j-bowhay commented:
Thanks @math-hiyoko, would you be able to change the `use_abs` to what is suggested in https://github.com/scipy/scipy/issues/22685#issuecomment-2813538308

math-hiyoko commented:
Just to confirm—I’ll add a new variant parameter to sparse.csgraph.laplacian, defaulting to "default", and "use_abs" will activate the functionality implemented in this PR.

math-hiyoko committed:
ENH: parameter: use_abs -> variant

math-hiyoko committed:
FIX: doc format

math-hiyoko committed:
FIX: benchmark parameter

math-hiyoko commented:
I’ve updated the PR to replace the boolean use_abs parameter with the new variant parameter

dschult commented on scipy/sparse/csgraph/tests/test_graph_laplacian.py:
```diff
@@ -176,42 +184,71 @@ def _check_laplacian_dtype(
-    A = [[0, 1, 0],
-         [4, 2, 0],
-         [0, 0, 0]]
+    A = [[0, 2, -2],
```
This is changing the tests that were already in place. Is there a reason to do that?

math-hiyoko replied:
The `use_abs` variant specifically addresses cases where edge weights can be negative. The original test cases did not include negative edge weights, making them unsuitable for testing the behavior of the `use_abs` variant.

math-hiyoko replied:
This PR also adds the method="use_abs" case to this test. Now both cases are being tested. 
https://github.com/math-hiyoko/scipy/blob/bf8bcb89b7020028f753398cae2f36a2b2d451ad/scipy/sparse/csgraph/tests/test_graph_laplacian.py#L225-L251

dschult replied:
Maybe I'm not understanding something. The test is filled with if statements for each case and sets matrices `A` and vector `d`.  The cases that have `method="default"` dont have anything to do with `use_abs`.

In fact, maybe all the different cases could be separated nicely using pytest fixtures. But that's not a blocker for this PR.

I would like that the existing tests are still run (unless there is a reason they shouldn't be). Adding new ones is fine. But I'd prefer not to take remove existing tests.

math-hiyoko replied:
Understood—I'll restore the original test cases and have them coexist alongside the new tests introduced in this PR.

dschult reviewed (commented):
Thanks for this @math-hiyoko 

I'd like other folks take on having the string `"default"` be the default variant. We could use `None`, or `"default"` or `"degree"` (or some other descriptive name). It seems like the string "default" doesn't convey a meaning -- beyond the fact that it is being used as the default variant -- which we already know because it is in the function signature. Using `None` provides easy/quick look-up with `variant is None`. And a descriptive word provides meaning and allows potential sqitching of the default without switching the name of the default variant.  I think `"degree"` is best of these 3 options, but maybe there is a better one-word description of what the standard laplacian uses to fill the diagonal.

My second comment is that the changes to the tests add new tests, but also change the existing tests (see comment below). Is there a reason you chose to change the current arrays that are tested? (Are these better for the original laplacian as well as for the new variant we are introducing?)

math-hiyoko committed:
FIX: restore original testcase

math-hiyoko committed:
default -> degree

dschult reviewed (commented):
- We need the docs to have a reference for Kunegis et al (2010).
- We should get a better name than `use_abs`. What should this variant be called?
- see below for comments about duplicate/unneeded computation.

Can you check that the formula is right for this `use_abs` variant? 
I thought it should be L = |A|.sum() - |A|   but you have it as L = |A|.sum() - A.
The latter is strange for a laplacian because the rows don't sum to zero (that is the usual eigenvector of all ones is not an eigenvector). That is a fairly universal property of matrices that are called laplacians.

fernandodiazdiaz commented:
Perhaps I can be of some assistance with the mathematical details. I can confirm that the correct expression is L=|A|.sum() - A. Indeed, this Laplacian has the unusual property that its columns do not add up to one. This is the price to pay to have a positive semidefinite operator. There are some authors that use instead L=A.sum() - A. This alternative choice has rows that add up to zero, but it ceases to be positive semidefinite, and therefore the associated dynamical systems may exhibit unstable behaviors. When both Laplacians are compared, the first one is typically called the "opposing Laplacian", while the second one is the "repelling Laplacian". Let me emphasize that the object L=|A|.sum() - |A| is another Laplacian that is not suited for the analysis of signed networks, as it removes all the information contained in the sign. I typically refer to it as the "unsigned Laplacian", although there is not a single name for it in the literature. 

A more detailed discussion on the reasoning behind each type of Laplacian can be found here : Shi, Guodong, Claudio Altafini, and John S. Baras. "Dynamics over signed networks." Siam review 61.2 (2019): 229-257.

dschult commented:
Thanks for this comment @fernandodiazdiaz 
Do you have insight into when one might use the laplacian variant `A.sum() + A`? Or the normalized laplacians like: `I - (1 / np.sqrt(A.sum()) * A * (1 / np.sqrt(A.sum()) )`?  There seem to be so many different variants!

This says to me that we definitely need a better name for the variant than `use_abs` -- other laplacian definitions use abs is other ways.  What is a good name for this variant?

fernandodiazdiaz commented:
The Laplacian variant A.sum() + A is typically referred to as the "signless Laplacian". Paradoxically, its main application is when A is unsigned, where it is used to get some insight into the system's bipartivity. For example, a network is bipartite if and only if the singless Laplacian is singular, and the eigenvector corresponding to the null eigenvalue provides the partition of the node set that makes all edges fall between nodes in different subsets. 

With respect to the normalized Laplacians, their use is completely analogous to that of the unsigned case. For instance, one can define a random walk opposing Laplacian L_rw = I-|A|.sum()^{-1}A, which governs the behavior of a random walk on a signed network. When using these normalizations, the degree matrix must be computed using the absolute value; otherwise one risks divisions by zero. 

I agree that 'use_abs' may be too vague for some users, as different Laplacians use the absolute value. A more concrete option could be 'type': {'opposing', 'repelling', 'unsigned'}, where 'unsigned' corresponds to the default.

math-hiyoko committed:
DOC: add reference to Kunegis et al (2010)

math-hiyoko committed:
DOC: fix reference

dschult commented:
The normalized cases are analogous to the unnormalized cases. But do we want this function to be able to provide the normalized cases too by choosing a variant? Or is it better to have this function only  provide unnormalized cases and have the user create normalized versions from the unnormalized results. Your example of `I-|A|.sum()^{-1}A` makes it seem like the user could create this normalized version from the unnormalized by multiplying by `|A|.sum()^{-1}`. And I guess the example `I - A.sum()^(-1/2) * A * A.sum()^(-1/2)` would require pre and post multiplication by `D=A.sum()^(-1/2)`. 

Are there normalized versions that won't be easily created from the unnormalized versions?  I'm just trying to get a feel for how many variants we'll need in the long term.

math-hiyoko committed:
FIX: lint

math-hiyoko committed:
ENH: avoid multiple summing

math-hiyoko commented:
> - We need the docs to have a reference for Kunegis et al (2010).
> - We should get a better name than use_abs. What should this variant be called?
> - see below for comments about duplicate/unneeded computation.
>
> Can you check that the formula is right for this use_abs variant?
> I thought it should be L = |A|.sum() - |A| but you have it as L = |A|.sum() - A.
> The latter is strange for a laplacian because the rows don't sum to zero (that is the usual eigenvector of all ones is not an eigenvector). That is a fairly universal property of matrices that are called laplacians.

@dschult 

Thanks for the comments!
I’ve addressed the first and third points:
- Added the reference to Kunegis et al. (2010) in the docs
- Removed the duplicate/unnecessary computation based on the variant setting.

math-hiyoko commented:
@fernandodiazdiaz @dschult 
As part of this PR, would it make sense to rename the current default variant to "unsigned" and use_abs to "opposing", and additionally introduce a new "repelling" variant?
Please let me know if this direction sounds good.

math-hiyoko committed:
fix: degree -> unsigned, use_abs -> opposing

math-hiyoko committed:
fix: implemented unsigned

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko commented:
@fernandodiazdiaz @dschult
I’ve gone through Shi, Altafini & Baras (2019) “Dynamics over Signed Networks” to make sure I fully understood the mathematical distinctions.
From that reading it’s clear that:
- the Laplacian already in main is the repelling variant (L = D – A, degrees from signed weights), so changing its behaviour would break existing code;

I’ve therefore left the current implementation as variant="repelling" for backward-compatibility and added fully-tested implementations of "opposing" and "unsigned", plus docstrings, references and unit tests for each.

Let me know if anything should be tweaked further!

math-hiyoko committed:
fix: lint

math-hiyoko committed:
fix: test variant only for negative edge test

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko committed:
fix: docstring

math-hiyoko committed:
Merge branch 'feature/#22685' of github.com:math-hiyoko/scipy into feature/#22685

math-hiyoko committed:
fix: copy behavior

math-hiyoko commented:
Addresses an environment-dependent bug.

dschult reviewed (commented):
For the "unsigned" variant we replace the edge weights with their absolute value. And in some places, the PR overwrites the values in the graph object to avoid memory allocation `np.abs(m.data, out=m.data)`.  Does this change the input graph? Should we warn users about this? Or do we make a copy before the code ever gets to this spot?

I think this is a reasonable approach for capturing all the variants we have discussed.
:)

math-hiyoko committed:
Update scipy/sparse/csgraph/_laplacian.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko commented:
> For the "unsigned" variant we replace the edge weights with their absolute value. And in some places, the PR overwrites the values in the graph object to avoid memory allocation np.abs(m.data, out=m.data). Does this change the input graph? Should we warn users about this? Or do we make a copy before the code ever gets to this spot?

With the current implementation, we do not overwrite the input graph’s data in-place—the original adjacency matrix remains unchanged. This behavior is explicitly verified by the _check_symmetric_graph_laplacian helper (exercised via test_symmetric_graph_laplacian and test_sparse_formats), which asserts that when `copy=True` the input stays intact.
https://github.com/math-hiyoko/scipy/blob/5bcfdcd47a23c510ae00641ab15a7065f945a9fa/scipy/sparse/csgraph/tests/test_graph_laplacian.py#L68-L75

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko committed:
fix: test laplacian for negative edge graphs

math-hiyoko committed:
fix: fix test for the cases variant=opposing

math-hiyoko committed:
fix: the case copy='True' and variant='unsigned'

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko commented:
Thank you for catching this.
I’ve updated the tests to validate the behavior of the copy parameter when edge weights are negative, and I’ve also corrected the laplacian implementation accordingly.

math-hiyoko committed:
fix: the case copy='False' and variant='unsigned'

math-hiyoko commented:
Thank you for catching this.
I’ve updated the tests to validate the behavior of the copy parameter when edge weights are negative, and I’ve also corrected the laplacian implementation accordingly.

dschult commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -69,6 +70,20 @@ def laplacian(
+        Detailed discussions on the reasoning behind each type of variant
+        can be found in [2].:
+
+        * 'repelling' is standard degree calculation using edge weights.
+        * 'opposing' computes the degree using absolute values of edge weights.
+          This prevents cancellation of positive and negative edge weights in
+          signed graphs, as described in [3].
+        * 'unsigned' discards the sign of edge weights.
+          This removes all the information contained in the sign.
```
This description talks about calculation of degree values. But not the sign used in the definition of L itself. I think this will be confusing for someone coming to this function who has never heard of or cares about signed graphs. 

Do I understand correctly that in addition to the treatment of signs when computing degree, that some of these variants compute `L=D-A` and others compute `L=D+A`?  And maybe it would help if this description started with some indication that this keyword handles variants for signed graphs. That will help folks who are using standard edge weights figure out that they don't need to worry about this keyword arg.

dschult commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -271,6 +298,57 @@ def laplacian(
+    array([2, 2, 2])
+
+    Setting variant="unsigned" discards the sign of edge weights
+    and computes the Laplacian using absolute edge weights
```
It's not clear in this sentence what "computes the Laplacian using absolute edge weights" refers to. Knowing the answer I must guess that "computes the Laplacian" refers to handling the adjacency matrix part of the laplacian formula. But computing the laplacian involves both computing degrees and combining the degrees with the adjacency info. Let's either find better words to use, use formulas, or define terms for the degree handling and for the adjacency handling and use them.

dschult reviewed (commented):
I think the words "repelling" "opposing" and "unsigned" are all from the signed graph literature. Most people coming to this function will not be using signed graphs. So we want a clear path for them to use/ignore the variant keyword.  One way to do this would be to rename the keyword to something like: `signed_graph_variant`. That's long and ugly but contains the right information.  We could instead leave the name `variant` and make the default name be something like `standard` and use the text to explain that the signed graph literature calls this `repelling`. 

We also should make sure each description (in keyword description and examples) explains both the treatment of signs in degree and treatment of the sign of adjacency matrix in the formula `L = D - A`. We either need to define the words we are using or use formulas. 

In the end I'd like the signed graph part of the laplacian function to clearly be a special case, and have the words used to describe the options be clearly defined and complete in their description.

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko committed:
fix: variant -> signed_graph_variant

math-hiyoko committed:
enh: repelling -> standard

math-hiyoko committed:
fix: docstring

math-hiyoko committed:
fix: docstring

math-hiyoko committed:
fix: lint

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko committed:
fix: reviewed docs

math-hiyoko commented:
@dschult 
Thank you for the feedback!
Based on your suggestions, I’ve renamed the parameter to signed_graph_variant (defaulting to standard), updated the docstring to include the explicit formulas for each variant, and verified the changes by running spin docs.

dschult reviewed (approved):
Thanks @math-hiyoko For the changes here.
This looks ready to be merged to me.

It keeps the current features of `csgraph.laplacian` and adds features for signed graphs. The docs are clear that for people not using signed graphs, they can just ignore that keyword and for people using signed graphs they can see the options described clearly with citations for more information. The additional tests look good.

I approve this PR.

j-bowhay reviewed (commented):
Thanks, some minor comments

j-bowhay commented:
Also CI failure is related

math-hiyoko committed:
Apply suggestions from code review

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

math-hiyoko committed:
enh: test error msg

math-hiyoko committed:
Merge branch 'main' into feature/#22685

j-bowhay commented:
CI failures look related, do you mind taking a look

math-hiyoko committed:
fix: lint error

math-hiyoko committed:
Merge branch 'feature/#22685' of github.com:math-hiyoko/scipy into feature/#22685

math-hiyoko committed:
Merge branch 'main' into feature/#22685

dschult committed:
attempt to fix sphinx formatting of nested lists

math-hiyoko commented:
The failing test is the issue of #23412

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko committed:
Merge branch 'main' into feature/#22685

j-bowhay committed:
TST: complete test coverage

j-bowhay committed:
STY: minor style nit
[skip ci]

j-bowhay reviewed (changes requested):
Sorry for drawing this out even longer but there seems to be a slight issue, the function is documented to work with sparse arrays however the new argument can break this:
```
In [20]: G = np.array([[0, 1, -1], [1, 0,  1], [-1, 1, 0]])

In [21]: G = sparse.dok_array(G)

In [22]: laplacian(G, form="lo")
Out[22]: <3x3 _CustomLinearOperator with dtype=int64>

In [23]: laplacian(G, signed_graph_variant="unsigned", form="lo")
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[23], line 1
----> 1 laplacian(G, signed_graph_variant="unsigned", form="lo")

File ~/development/scipy/scipy/sparse/csgraph/_laplacian.py:453, in laplacian(csgraph, normed, return_diag, use_out_degree, copy, form, dtype, symmetrized, signed_graph_variant)
    445     create_lap = (
    446         _laplacian_sparse_flo
    447         if issparse(csgraph)
    448         else _laplacian_dense_flo
    449     )
    451 degree_axis = 1 if use_out_degree else 0
--> 453 lap, d = create_lap(
    454     csgraph,
    455     normed=normed,
    456     axis=degree_axis,
    457     copy=copy,
    458     form=form,
    459     dtype=dtype,
    460     symmetrized=symmetrized,
    461     signed_graph_variant=signed_graph_variant,
    462 )
    463 if is_pydata_sparse:
    464     lap = pydata_sparse_cls.from_scipy_sparse(lap)

File ~/development/scipy/scipy/sparse/csgraph/_laplacian.py:513, in _laplacian_sparse_flo(***failed resolving arguments***)
    509     dtype = graph.dtype
    511 if signed_graph_variant == "unsigned":
    512     # Discard the sign of edge weights.
--> 513     np.abs(graph.data, out=graph.data)
    515 graph_sum = np.asarray(graph.sum(axis=axis)).ravel()
    516 if symmetrized:

AttributeError: 'dok_array' object has no attribute 'data'
```

math-hiyoko committed:
fix: laplacian_sparse_flo for dok and lil

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko commented:
Hi @j-bowhay,  I’ve made the requested changes.
Could you take another look when you get a chance? Thanks!

math-hiyoko committed:
Merge branch 'main' into feature/#22685

j-bowhay commented:
Apologies @math-hiyoko for the delay, I still need to find the time to look at this carefully.

math-hiyoko committed:
Merge branch 'main' into feature/#22685

dschult committed:
Merge branch 'main' into feature/#22685

dschult commented:
I merged with main again to pull in some recent fixes in main.
Still some failures apparently unrelated to this PR. @math-hiyoko I just wanted to warn you that you should pull from upstream main to bring this merge to your local repo. :)

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko committed:
Merge branch 'main' into feature/#22685

math-hiyoko committed:
Merge branch 'main' into feature/#22685

lucascolley committed:
Merge branch 'main' into feature/#22685

tylerjereddy commented:
Looks like a core dev still wants to do a detailed review here per https://github.com/scipy/scipy/pull/22830#issuecomment-3288875235 and it seems unlikely this will cross the line before we branch in a day or two, so I'll bump the milestone.

math-hiyoko committed:
Merge branch 'main' into feature/#22685

tylerjereddy commented:
The concern from https://github.com/scipy/scipy/pull/22830#issuecomment-3619181816 is still valid and this hasn't seen activity apart from merging in `main` in the last six months, so I'll bump the milestone with branching in ~10 days (the PR is still "only" a year old, so maybe there's still some hope to get a maintainer to take a closer look).
